Make user e-mail field case insensitive to prevent multiple account with same email

sandeepsajan0 opened the pull request:
Fixes #2985

sandeepsajan0 committed:
Fix login issue for accounts with insensitive emails

sandeepsajan0 committed:
Restrict user to create via insensitive emails

sandeepsajan0 committed:
Add comments and error handling to user creation

theskumar commented on hypha/apply/users/models.py:
```diff
@@ -87,18 +93,49 @@ def create_superuser(self, email, password, **extra_fields):
         defaults.update(password=temp_pass_hash)
-        user, created = self.get_or_create(defaults=defaults, **kwargs)
-        if created:
+        user = self.filter(email__iexact=kwargs.get('email')).first()  # case insensitive matching
```
use `get_user_by_email` here?

sandeepsajan0 replied:
At the time of user creation, we only need a case-insensitive search.  We can also use `get_user_by_email` but for that, we will need to pass an extra argument and will need to update the logic of the function a bit.  But yes, our logic will only be in one place. Let me try with that logic and see if it looks good.

theskumar replied:
Let's not do this

theskumar commented on hypha/apply/users/models.py:
```diff
@@ -66,6 +69,9 @@ def _create_user(self, email, password, **extra_fields):
         if not email:
             raise ValueError('The given email must be set')
         email = self.normalize_email(email)
+        existing_user = self.filter(email__iexact=email)
```
use `get_user_by_email` here?

theskumar commented on hypha/apply/users/utils.py:
```diff
@@ -7,6 +8,21 @@
 from django.utils.http import urlsafe_base64_encode
 
 
+def get_user_by_email(email):
```
Please add test for this function.

theskumar commented on hypha/apply/users/models.py:
```diff
@@ -87,18 +93,49 @@ def create_superuser(self, email, password, **extra_fields):
 
         return self._create_user(email, password, **extra_fields)
 
+    def _extract_model_params(self, defaults, **kwargs):
```
I"m curious to know why this function needs to be implemented or overridden.

sandeepsajan0 replied:
It is not overridden, we just copied it from Queryset class(as mentioned in the comment). 
It was inaccessible directly, maybe because of the custom manager(`from_queryset`).

sandeepsajan0 committed:
Update get_user_by_email logic, make case sensitive search optional

sandeepsajan0 committed:
Add test for get_user_by_email method

theskumar commented on hypha/apply/users/models.py:
```diff
@@ -91,18 +97,49 @@ def create_superuser(self, email, password, **extra_fields):
         defaults.update(password=temp_pass_hash)
-        user, created = self.get_or_create(defaults=defaults, **kwargs)
-        if created:
+        user = get_user_by_email(kwargs.get('email'), sensitive_search=0)  # case insensitive matching
```
Please not do this here. I'll provide solution here.

theskumar committed:
Handle multiple user accounts exist during submissions

- create is_user_already_registered func
- raise error during application submissions when
  not logged in if user provides an email that is 
  present in multiple account.

theskumar committed:
update docstring for User -> get_or_create_and_notify

theskumar committed:
add/update test cases

theskumar reviewed (changes requested):
Please ensure applicant group is added to the user if they have atleast one application submisison

theskumar committed:
fix isort

sandeepsajan0 committed:
Add error message to submission form

fourthletter commented:
Amazing! Thank you 💯 

![image](https://user-images.githubusercontent.com/20019656/193823657-5dc8333c-764b-488a-af26-67bc1edd2fd9.png)

sandeepsajan0 commented on hypha/apply/stream_forms/forms.py:
```diff
@@ -41,8 +47,35 @@ def hidden_fields(self):
-    pass
+    """ Adds page and user reference to the form class"""
+
+    def clean(self):
```
@frjo I checked this piece of code is not getting used by any review forms or determination forms. Both(Review and determination) forms have their own form classes. So, I think we can move forward with this change. What do you think?

theskumar reviewed (approved):
@sandeepsajan0 feel free to remove https://github.com/HyphaApp/hypha/pull/2989#discussion_r987862931 or open a new PR just for removing the unused form.
